File: src/dere_daemon/routers/core_memory.py
"""Core memory block endpoints."""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from dere_daemon.dependencies import get_db
from dere_shared.models import ConsolidationRun, CoreMemoryBlock, CoreMemoryVersion, Session

router = APIRouter(prefix="/memory", tags=["core_memory"])

CORE_MEMORY_BLOCK_TYPES = {"persona", "human", "task"}
DEFAULT_CHAR_LIMIT = 8192


class CoreMemoryEditRequest(BaseModel):
    block_type: Literal["persona", "human", "task"]
    content: str
    reason: str | None = None
    scope: Literal["user", "session"] = "user"
    session_id: int | None = None
    user_id: str | None = None
    char_limit: int | None = None


class CoreMemoryBlockResponse(BaseModel):
    id: int
    block_type: str
    content: str
    scope: str
    session_id: int | None
    user_id: str | None
    char_limit: int
    version: int
    updated_at: datetime | None


class CoreMemoryEditResponse(BaseModel):
    block: CoreMemoryBlockResponse
    created: bool


class CoreMemoryHistoryResponse(BaseModel):
    block_id: int
    version: int
    content: str
    reason: str | None
    created_at: datetime | None


class CoreMemoryRollbackRequest(BaseModel):
    block_type: Literal["persona", "human", "task"]
    target_version: int
    reason: str | None = None
    scope: Literal["user", "session"] = "user"
    session_id: int | None = None
    user_id: str | None = None


class CoreMemoryRollbackResponse(BaseModel):
    block: CoreMemoryBlockResponse
    rolled_back_to: int


class ConsolidationRunResponse(BaseModel):
    id: int
    user_id: str | None
    task_id: int | None
    status: str
    started_at: datetime | None
    finished_at: datetime | None
    recency_days: int | None
    community_resolution: float | None
    update_core_memory: bool
    triggered_by: str | None
    stats: dict | None
    error_message: str | None


class ConsolidationRunListResponse(BaseModel):
    runs: list[ConsolidationRunResponse]
    total: int
    offset: int
    limit: int


def _validate_content(content: str, char_limit: int) -> None:
    if char_limit <= 0:
        raise HTTPException(status_code=400, detail="char_limit must be positive")
    if len(content) > char_limit:
        raise HTTPException(
            status_code=400,
            detail=f"Content exceeds char_limit ({len(content)}/{char_limit})",
        )


def _block_response(block: CoreMemoryBlock, scope: str) -> CoreMemoryBlockResponse:
    return CoreMemoryBlockResponse(
        id=block.id or 0,
        block_type=block.block_type,
        content=block.content,
        scope=scope,
        session_id=block.session_id,
        user_id=block.user_id,
        char_limit=block.char_limit,
        version=block.version,
        updated_at=block.updated_at,
    )


def _run_response(run: ConsolidationRun) -> ConsolidationRunResponse:
    return ConsolidationRunResponse(
        id=run.id or 0,
        user_id=run.user_id,
        task_id=run.task_id,
        status=run.status,
        started_at=run.started_at,
        finished_at=run.finished_at,
        recency_days=run.recency_days,
        community_resolution=run.community_resolution,
        update_core_memory=run.update_core_memory,
        triggered_by=run.triggered_by,
        stats=run.stats,
        error_message=run.error_message,
    )


async def _resolve_block(
    db: AsyncSession,
    *,
    block_type: str,
    scope: str,
    session_id: int | None,
    user_id: str | None,
) -> tuple[CoreMemoryBlock | None, str]:
    resolved_scope = scope
    if scope == "session":
        if session_id is None:
            raise HTTPException(status_code=400, detail="session_id required for session scope")
        stmt = select(CoreMemoryBlock).where(
            CoreMemoryBlock.session_id == session_id,
            CoreMemoryBlock.block_type == block_type,
        )
        result = await db.execute(stmt)
        return result.scalar_one_or_none(), resolved_scope

    resolved_user_id = user_id
    if resolved_user_id is None and session_id is not None:
        session = await db.get(Session, session_id)
        resolved_user_id = session.user_id if session else None
    if not resolved_user_id:
        raise HTTPException(status_code=400, detail="user_id not found for session")

    stmt = select(CoreMemoryBlock).where(
        CoreMemoryBlock.user_id == resolved_user_id,
        CoreMemoryBlock.session_id.is_(None),
        CoreMemoryBlock.block_type == block_type,
    )
    result = await db.execute(stmt)
    return result.scalar_one_or_none(), resolved_scope


@router.post("/core/edit", response_model=CoreMemoryEditResponse)
async def edit_core_memory(
    payload: CoreMemoryEditRequest,
    request: Request,
    db: AsyncSession = Depends(get_db),
):
    """Create or update a core memory block."""
    block_type = payload.block_type.strip().lower()
    if block_type not in CORE_MEMORY_BLOCK_TYPES:
        raise HTTPException(status_code=400, detail="Invalid block_type")

    scope = payload.scope
    session_id = payload.session_id
    user_id = payload.user_id

    if scope == "session":
        if session_id is None:
            raise HTTPException(status_code=400, detail="session_id required for session scope")
    else:
        if user_id is None and session_id is None:
            raise HTTPException(status_code=400, detail="user_id or session_id required")
        if user_id is None and session_id is not None:
            session = await db.get(Session, session_id)
            user_id = session.user_id if session else None
        if not user_id:
            raise HTTPException(status_code=400, detail="user_id not found for session")

    if payload.char_limit is not None and payload.char_limit <= 0:
        raise HTTPException(status_code=400, detail="char_limit must be positive")

    now = datetime.now(UTC)

    block, resolved_scope = await _resolve_block(
        db,
        block_type=block_type,
        scope=scope,
        session_id=session_id,
        user_id=user_id,
    )

    created = False
    if block:
        char_limit = payload.char_limit or block.char_limit or DEFAULT_CHAR_LIMIT
        _validate_content(payload.content, char_limit)
        block.content = payload.content
        block.char_limit = char_limit
        block.version = (block.version or 1) + 1
        block.updated_at = now
    else:
        created = True
        char_limit = payload.char_limit or DEFAULT_CHAR_LIMIT
        _validate_content(payload.content, char_limit)
        block = CoreMemoryBlock(
            user_id=None if scope == "session" else user_id,
            session_id=session_id if scope == "session" else None,
            block_type=block_type,
            content=payload.content,
            char_limit=char_limit,
            version=1,
            created_at=now,
            updated_at=now,
        )
        db.add(block)
        await db.flush()

    version = CoreMemoryVersion(
        block_id=block.id,  # type: ignore[arg-type]
        version=block.version,
        content=block.content,
        reason=payload.reason,
        created_at=now,
    )
    db.add(version)
    await db.commit()

    return CoreMemoryEditResponse(block=_block_response(block, resolved_scope), created=created)


@router.get("/core", response_model=list[CoreMemoryBlockResponse])
async def list_core_memory(
    request: Request,
    session_id: int | None = None,
    user_id: str | None = None,
    db: AsyncSession = Depends(get_db),
):
    """List core memory blocks for a session or user."""
    resolved_user_id = user_id
    if session_id is not None:
        session = await db.get(Session, session_id)
        if session and session.user_id:
            resolved_user_id = session.user_id

    if not session_id and not resolved_user_id:
        raise HTTPException(status_code=400, detail="session_id or user_id required")

    blocks: dict[str, CoreMemoryBlock] = {}
    if session_id is not None:
        stmt = select(CoreMemoryBlock).where(
            CoreMemoryBlock.session_id == session_id,
            CoreMemoryBlock.block_type.in_(CORE_MEMORY_BLOCK_TYPES),
        )
        result = await db.execute(stmt)
        for block in result.scalars().all():
            blocks[block.block_type] = block

    if resolved_user_id:
        stmt = select(CoreMemoryBlock).where(
            CoreMemoryBlock.user_id == resolved_user_id,
            CoreMemoryBlock.session_id.is_(None),
            CoreMemoryBlock.block_type.in_(CORE_MEMORY_BLOCK_TYPES),
        )
        result = await db.execute(stmt)
        for block in result.scalars().all():
            if block.block_type not in blocks:
                blocks[block.block_type] = block

    response = []
    for block_type in ("persona", "human", "task"):
        block = blocks.get(block_type)
        if not block:
            continue
        scope = "session" if block.session_id else "user"
        response.append(_block_response(block, scope))

    return response


@router.get("/core/history", response_model=list[CoreMemoryHistoryResponse])
async def core_memory_history(
    request: Request,
    block_type: Literal["persona", "human", "task"],
    limit: int = 20,
    scope: Literal["user", "session"] = "user",
    session_id: int | None = None,
    user_id: str | None = None,
    db: AsyncSession = Depends(get_db),
):
    """Return version history for a core memory block."""
    block, _scope = await _resolve_block(
        db,
        block_type=block_type,
        scope=scope,
        session_id=session_id,
        user_id=user_id,
    )
    if not block:
        raise HTTPException(status_code=404, detail="Core memory block not found")

    stmt = (
        select(CoreMemoryVersion)
        .where(CoreMemoryVersion.block_id == block.id)
        .order_by(CoreMemoryVersion.version.desc())
        .limit(limit)
    )
    result = await db.execute(stmt)
    versions = result.scalars().all()

    return [
        CoreMemoryHistoryResponse(
            block_id=block.id or 0,
            version=version.version,
            content=version.content,
            reason=version.reason,
            created_at=version.created_at,
        )
        for version in versions
    ]


@router.post("/core/rollback", response_model=CoreMemoryRollbackResponse)
async def core_memory_rollback(
    payload: CoreMemoryRollbackRequest,
    request: Request,
    db: AsyncSession = Depends(get_db),
):
    """Rollback a core memory block to a previous version."""
    block, resolved_scope = await _resolve_block(
        db,
        block_type=payload.block_type,
        scope=payload.scope,
        session_id=payload.session_id,
        user_id=payload.user_id,
    )
    if not block:
        raise HTTPException(status_code=404, detail="Core memory block not found")

    if payload.target_version <= 0:
        raise HTTPException(status_code=400, detail="target_version must be positive")

    stmt = select(CoreMemoryVersion).where(
        CoreMemoryVersion.block_id == block.id,
        CoreMemoryVersion.version == payload.target_version,
    )
    result = await db.execute(stmt)
    target = result.scalar_one_or_none()
    if not target:
        raise HTTPException(status_code=404, detail="Target version not found")

    _validate_content(target.content, block.char_limit)

    now = datetime.now(UTC)
    block.content = target.content
    block.version = (block.version or 1) + 1
    block.updated_at = now

    reason = payload.reason or f"rollback to v{payload.target_version}"
    db.add(
        CoreMemoryVersion(
            block_id=block.id,  # type: ignore[arg-type]
            version=block.version,
            content=block.content,
            reason=reason,
            created_at=now,
        )
    )
    await db.commit()

    return CoreMemoryRollbackResponse(
        block=_block_response(block, resolved_scope),
        rolled_back_to=payload.target_version,
    )


@router.get("/consolidation/runs", response_model=ConsolidationRunListResponse)
async def list_consolidation_runs(
    request: Request,
    user_id: str | None = None,
    status: str | None = None,
    limit: int = 20,
    offset: int = 0,
    db: AsyncSession = Depends(get_db),
):
    """List recent memory consolidation runs."""
    limit = max(1, min(limit, 100))
    offset = max(0, offset)

    stmt = select(ConsolidationRun)
    count_stmt = select(func.count()).select_from(ConsolidationRun)

    if user_id:
        stmt = stmt.where(ConsolidationRun.user_id == user_id)
        count_stmt = count_stmt.where(ConsolidationRun.user_id == user_id)
    if status:
        stmt = stmt.where(ConsolidationRun.status == status)
        count_stmt = count_stmt.where(ConsolidationRun.status == status)

    total = (await db.execute(count_stmt)).scalar() or 0
    result = await db.execute(
        stmt.order_by(ConsolidationRun.started_at.desc()).offset(offset).limit(limit)
    )
    runs = result.scalars().all()

    return ConsolidationRunListResponse(
        runs=[_run_response(run) for run in runs],
        total=total,
        offset=offset,
        limit=limit,
    )

File: packages/daemon/src/core-memory.ts
import type { Hono } from "hono";

import { getDb } from "./db.js";

const CORE_MEMORY_BLOCK_TYPES = new Set(["persona", "human", "task"]);
const DEFAULT_CHAR_LIMIT = 8192;

type CoreMemoryBlockRow = {
  id: number;
  block_type: string;
  content: string;
  scope: "user" | "session";
  session_id: number | null;
  user_id: string | null;
  char_limit: number;
  version: number;
  updated_at: Date | null;
};

type ConsolidationRunRow = {
  id: number;
  user_id: string | null;
  task_id: number | null;
  status: string;
  started_at: Date | null;
  finished_at: Date | null;
  recency_days: number | null;
  community_resolution: number | null;
  update_core_memory: boolean;
  triggered_by: string | null;
  stats: Record<string, unknown> | null;
  error_message: string | null;
};

function nowDate(): Date {
  return new Date();
}

async function parseJson<T>(req: Request): Promise<T | null> {
  try {
    return (await req.json()) as T;
  } catch {
    return null;
  }
}

function toNumber(value: string | undefined, fallback: number): number {
  if (value === undefined || value === null || value === "") {
    return fallback;
  }
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : fallback;
}

function parseNullableNumber(value: string | undefined): number | null {
  if (value === undefined || value === null || value === "") {
    return null;
  }
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : null;
}

function normalizeBlockType(value: unknown): string | null {
  if (typeof value !== "string") {
    return null;
  }
  const normalized = value.trim().toLowerCase();
  return CORE_MEMORY_BLOCK_TYPES.has(normalized) ? normalized : null;
}

function validateContent(content: string, charLimit: number): string | null {
  if (charLimit <= 0) {
    return "char_limit must be positive";
  }
  if (content.length > charLimit) {
    return `Content exceeds char_limit (${content.length}/${charLimit})`;
  }
  return null;
}

function toBlockResponse(block: CoreMemoryBlockRow) {
  return {
    id: block.id,
    block_type: block.block_type,
    content: block.content,
    scope: block.scope,
    session_id: block.session_id,
    user_id: block.user_id,
    char_limit: block.char_limit,
    version: block.version,
    updated_at: block.updated_at,
  };
}

function toRunResponse(run: ConsolidationRunRow) {
  return {
    id: run.id,
    user_id: run.user_id,
    task_id: run.task_id,
    status: run.status,
    started_at: run.started_at,
    finished_at: run.finished_at,
    recency_days: run.recency_days,
    community_resolution: run.community_resolution,
    update_core_memory: run.update_core_memory,
    triggered_by: run.triggered_by,
    stats: run.stats,
    error_message: run.error_message,
  };
}

async function resolveUserId(
  sessionId: number | null,
  userId: string | null,
): Promise<string | null> {
  if (userId) {
    return userId;
  }
  if (sessionId === null) {
    return null;
  }
  const db = await getDb();
  const session = await db
    .selectFrom("sessions")
    .select(["user_id"])
    .where("id", "=", sessionId)
    .executeTakeFirst();
  return session?.user_id ?? null;
}

async function resolveBlock(args: {
  blockType: string;
  scope: "user" | "session";
  sessionId: number | null;
  userId: string | null;
}) {
  const db = await getDb();
  if (args.scope === "session") {
    if (args.sessionId === null) {
      return null;
    }
    return db
      .selectFrom("core_memory_blocks")
      .selectAll()
      .where("session_id", "=", args.sessionId)
      .where("block_type", "=", args.blockType)
      .executeTakeFirst();
  }

  if (!args.userId) {
    return null;
  }

  return db
    .selectFrom("core_memory_blocks")
    .selectAll()
    .where("user_id", "=", args.userId)
    .where("session_id", "is", null)
    .where("block_type", "=", args.blockType)
    .executeTakeFirst();
}

export function registerCoreMemoryRoutes(app: Hono): void {
  app.post("/memory/core/edit", async (c) => {
    const payload = await parseJson<Record<string, unknown>>(c.req.raw);
    if (!payload || typeof payload !== "object") {
      return c.json({ error: "Invalid JSON payload" }, 400);
    }

    const blockType = normalizeBlockType(payload.block_type);
    if (!blockType) {
      return c.json({ error: "Invalid block_type" }, 400);
    }

    const content = payload.content;
    if (typeof content !== "string") {
      return c.json({ error: "content is required" }, 400);
    }

    const scope = payload.scope === "session" ? "session" : "user";
    const sessionId = typeof payload.session_id === "number" ? payload.session_id : null;
    let userId = typeof payload.user_id === "string" ? payload.user_id : null;
    const charLimit = typeof payload.char_limit === "number" ? payload.char_limit : null;
    const reason = typeof payload.reason === "string" ? payload.reason : null;

    if (charLimit !== null && charLimit <= 0) {
      return c.json({ error: "char_limit must be positive" }, 400);
    }

    if (scope === "session") {
      if (sessionId === null) {
        return c.json({ error: "session_id required for session scope" }, 400);
      }
    } else {
      if (!userId && sessionId === null) {
        return c.json({ error: "user_id or session_id required" }, 400);
      }
      if (!userId && sessionId !== null) {
        userId = await resolveUserId(sessionId, null);
      }
      if (!userId) {
        return c.json({ error: "user_id not found for session" }, 400);
      }
    }

    const block = await resolveBlock({
      blockType,
      scope,
      sessionId,
      userId,
    });

    const now = nowDate();
    let created = false;
    let updatedBlock: CoreMemoryBlockRow;

    const db = await getDb();
    if (block) {
      const limit = charLimit ?? block.char_limit ?? DEFAULT_CHAR_LIMIT;
      const validationError = validateContent(content, limit);
      if (validationError) {
        return c.json({ error: validationError }, 400);
      }
      const nextVersion = (block.version ?? 1) + 1;
      await db
        .updateTable("core_memory_blocks")
        .set({
          content,
          char_limit: limit,
          version: nextVersion,
          updated_at: now,
        })
        .where("id", "=", block.id)
        .execute();

      updatedBlock = {
        id: block.id,
        block_type: block.block_type,
        content,
        scope,
        session_id: block.session_id,
        user_id: block.user_id,
        char_limit: limit,
        version: nextVersion,
        updated_at: now,
      };
    } else {
      created = true;
      const limit = charLimit ?? DEFAULT_CHAR_LIMIT;
      const validationError = validateContent(content, limit);
      if (validationError) {
        return c.json({ error: validationError }, 400);
      }

      const inserted = await db
        .insertInto("core_memory_blocks")
        .values({
          user_id: scope === "session" ? null : userId,
          session_id: scope === "session" ? sessionId : null,
          block_type: blockType,
          content,
          char_limit: limit,
          version: 1,
          created_at: now,
          updated_at: now,
        })
        .returningAll()
        .executeTakeFirstOrThrow();

      updatedBlock = {
        id: inserted.id,
        block_type: inserted.block_type,
        content: inserted.content,
        scope,
        session_id: inserted.session_id,
        user_id: inserted.user_id,
        char_limit: inserted.char_limit,
        version: inserted.version,
        updated_at: inserted.updated_at,
      };
    }

    await db
      .insertInto("core_memory_versions")
      .values({
        block_id: updatedBlock.id,
        version: updatedBlock.version,
        content: updatedBlock.content,
        reason,
        created_at: now,
      })
      .execute();

    return c.json({ block: toBlockResponse(updatedBlock), created });
  });

  app.get("/memory/core", async (c) => {
    const sessionId = parseNullableNumber(c.req.query("session_id"));
    const userIdParam = c.req.query("user_id");

    let resolvedUserId = typeof userIdParam === "string" ? userIdParam : null;
    if (sessionId !== null) {
      const sessionUserId = await resolveUserId(sessionId, null);
      if (sessionUserId) {
        resolvedUserId = sessionUserId;
      }
    }

    if (sessionId === null && !resolvedUserId) {
      return c.json({ error: "session_id or user_id required" }, 400);
    }

    const db = await getDb();
    const blocks = new Map<string, CoreMemoryBlockRow>();

    if (sessionId !== null) {
      const sessionBlocks = await db
        .selectFrom("core_memory_blocks")
        .selectAll()
        .where("session_id", "=", sessionId)
        .where("block_type", "in", Array.from(CORE_MEMORY_BLOCK_TYPES))
        .execute();

      for (const block of sessionBlocks) {
        blocks.set(block.block_type, {
          id: block.id,
          block_type: block.block_type,
          content: block.content,
          scope: "session",
          session_id: block.session_id,
          user_id: block.user_id,
          char_limit: block.char_limit,
          version: block.version,
          updated_at: block.updated_at,
        });
      }
    }

    if (resolvedUserId) {
      const userBlocks = await db
        .selectFrom("core_memory_blocks")
        .selectAll()
        .where("user_id", "=", resolvedUserId)
        .where("session_id", "is", null)
        .where("block_type", "in", Array.from(CORE_MEMORY_BLOCK_TYPES))
        .execute();

      for (const block of userBlocks) {
        if (!blocks.has(block.block_type)) {
          blocks.set(block.block_type, {
            id: block.id,
            block_type: block.block_type,
            content: block.content,
            scope: "user",
            session_id: block.session_id,
            user_id: block.user_id,
            char_limit: block.char_limit,
            version: block.version,
            updated_at: block.updated_at,
          });
        }
      }
    }

    const response: Array<ReturnType<typeof toBlockResponse>> = [];
    for (const blockType of ["persona", "human", "task"]) {
      const block = blocks.get(blockType);
      if (!block) {
        continue;
      }
      response.push(toBlockResponse(block));
    }

    return c.json(response);
  });

  app.get("/memory/core/history", async (c) => {
    const blockType = normalizeBlockType(c.req.query("block_type"));
    if (!blockType) {
      return c.json({ error: "Invalid block_type" }, 400);
    }

    const scope = c.req.query("scope") === "session" ? "session" : "user";
    const sessionId = parseNullableNumber(c.req.query("session_id"));
    let userId = c.req.query("user_id") ?? null;

    if (scope === "session") {
      if (sessionId === null) {
        return c.json({ error: "session_id required for session scope" }, 400);
      }
    } else {
      if (!userId && sessionId === null) {
        return c.json({ error: "user_id not found for session" }, 400);
      }
      if (!userId && sessionId !== null) {
        userId = await resolveUserId(sessionId, null);
      }
      if (!userId) {
        return c.json({ error: "user_id not found for session" }, 400);
      }
    }

    const block = await resolveBlock({
      blockType,
      scope,
      sessionId,
      userId,
    });
    if (!block) {
      return c.json({ error: "Core memory block not found" }, 404);
    }

    const limit = Math.max(1, toNumber(c.req.query("limit"), 20));
    const db = await getDb();
    const versions = await db
      .selectFrom("core_memory_versions")
      .select(["block_id", "version", "content", "reason", "created_at"])
      .where("block_id", "=", block.id)
      .orderBy("version", "desc")
      .limit(limit)
      .execute();

    return c.json(
      versions.map((version) => ({
        block_id: version.block_id,
        version: version.version,
        content: version.content,
        reason: version.reason,
        created_at: version.created_at,
      })),
    );
  });

  app.post("/memory/core/rollback", async (c) => {
    const payload = await parseJson<Record<string, unknown>>(c.req.raw);
    if (!payload || typeof payload !== "object") {
      return c.json({ error: "Invalid JSON payload" }, 400);
    }

    const blockType = normalizeBlockType(payload.block_type);
    if (!blockType) {
      return c.json({ error: "Invalid block_type" }, 400);
    }

    const targetVersion =
      typeof payload.target_version === "number" ? payload.target_version : null;
    if (!targetVersion || targetVersion <= 0) {
      return c.json({ error: "target_version must be positive" }, 400);
    }

    const scope = payload.scope === "session" ? "session" : "user";
    const sessionId = typeof payload.session_id === "number" ? payload.session_id : null;
    let userId = typeof payload.user_id === "string" ? payload.user_id : null;
    const reason = typeof payload.reason === "string" ? payload.reason : null;

    if (scope === "session") {
      if (sessionId === null) {
        return c.json({ error: "session_id required for session scope" }, 400);
      }
    } else {
      if (!userId && sessionId === null) {
        return c.json({ error: "user_id not found for session" }, 400);
      }
      if (!userId && sessionId !== null) {
        userId = await resolveUserId(sessionId, null);
      }
      if (!userId) {
        return c.json({ error: "user_id not found for session" }, 400);
      }
    }

    const block = await resolveBlock({
      blockType,
      scope,
      sessionId,
      userId,
    });
    if (!block) {
      return c.json({ error: "Core memory block not found" }, 404);
    }

    const db = await getDb();
    const target = await db
      .selectFrom("core_memory_versions")
      .selectAll()
      .where("block_id", "=", block.id)
      .where("version", "=", targetVersion)
      .executeTakeFirst();

    if (!target) {
      return c.json({ error: "Target version not found" }, 404);
    }

    const validationError = validateContent(target.content, block.char_limit);
    if (validationError) {
      return c.json({ error: validationError }, 400);
    }

    const now = nowDate();
    const nextVersion = (block.version ?? 1) + 1;
    await db
      .updateTable("core_memory_blocks")
      .set({
        content: target.content,
        version: nextVersion,
        updated_at: now,
      })
      .where("id", "=", block.id)
      .execute();

    await db
      .insertInto("core_memory_versions")
      .values({
        block_id: block.id,
        version: nextVersion,
        content: target.content,
        reason: reason ?? `rollback to v${targetVersion}`,
        created_at: now,
      })
      .execute();

    const updatedBlock: CoreMemoryBlockRow = {
      id: block.id,
      block_type: block.block_type,
      content: target.content,
      scope,
      session_id: block.session_id,
      user_id: block.user_id,
      char_limit: block.char_limit,
      version: nextVersion,
      updated_at: now,
    };

    return c.json({
      block: toBlockResponse(updatedBlock),
      rolled_back_to: targetVersion,
    });
  });

  app.get("/memory/consolidation/runs", async (c) => {
    const userId = c.req.query("user_id");
    const status = c.req.query("status");
    let limit = toNumber(c.req.query("limit"), 20);
    let offset = toNumber(c.req.query("offset"), 0);

    limit = Math.max(1, Math.min(limit, 100));
    offset = Math.max(0, offset);

    const db = await getDb();
    let query = db.selectFrom("consolidation_runs");
    let countQuery = db.selectFrom("consolidation_runs");

    if (userId) {
      query = query.where("user_id", "=", userId);
      countQuery = countQuery.where("user_id", "=", userId);
    }
    if (status) {
      query = query.where("status", "=", status);
      countQuery = countQuery.where("status", "=", status);
    }

    const count = await countQuery.select(db.fn.countAll().as("count")).executeTakeFirst();
    const total = Number(count?.count ?? 0);

    const runs = await query
      .selectAll()
      .orderBy("started_at", "desc")
      .offset(offset)
      .limit(limit)
      .execute();

    return c.json({
      runs: runs.map((run) =>
        toRunResponse({
          id: run.id,
          user_id: run.user_id,
          task_id: run.task_id,
          status: run.status,
          started_at: run.started_at,
          finished_at: run.finished_at,
          recency_days: run.recency_days,
          community_resolution: run.community_resolution,
          update_core_memory: run.update_core_memory,
          triggered_by: run.triggered_by,
          stats: run.stats as Record<string, unknown> | null,
          error_message: run.error_message,
        }),
      ),
      total,
      offset,
      limit,
    });
  });
}
